### PR TITLE
Allow CUDA EP enable or disable TunableOp via session options and environment variable

### PR DESCRIPTION
**Pull request by cloudhan**
This ports #13116 from ROCm EP to CUDA EP

### CODE REVIEW

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 1]
```diff
@@ -198,6 +198,40 @@ void CUDAExecutionProvider::PerThreadContext::IncrementRegularRunCountBeforeGrap
 }
 #endif
 
+std::optional<std::string> LoadEnv(const std::string& name, const std::unordered_set<std::string>& valid_values) {
```
Comment (zhangyaobit) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:201]: Is this code duplicated for rocm ep as well?

File: onnxruntime/python/onnxruntime_pybind_state_common.cc [hunk 2]
```diff
@@ -24,6 +24,8 @@ size_t gpu_mem_limit = std::numeric_limits<size_t>::max();
 OrtCudnnConvAlgoSearch cudnn_conv_algo_search = OrtCudnnConvAlgoSearchExhaustive;
 // TODO remove deprecated global config
 bool do_copy_in_default_stream = true;
+// TODO remove deprecated global config
```
Comment (zhangyaobit) [inline onnxruntime/python/onnxruntime_pybind_state_common.cc:27]: What is deprecated here?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 1 above)
Comment (cloudhan) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: Moved to `onnxruntime/core/platform/env_var_utils.h`

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 #include "core/common/inlined_containers.h"
 #include "core/providers/shared_library/provider_api.h"
+#include <core/platform/env_var_utils.h>
```
Comment (zhangyaobit) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:6]: Sort alphabetically?

Comment (abudup) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:6]: Also, why angle braces (`<>`) for this include?

**Review by abudup — APPROVED** [review]
Looks good, thanks!

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 4]
```diff
@@ -3,7 +3,7 @@
 
 #include "core/common/inlined_containers.h"
 #include "core/providers/shared_library/provider_api.h"
-#include <core/platform/env_var_utils.h>
+#include "core/platform/env_var_utils.h"
```
Comment (zhangyaobit) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:6]: move this line before "core/providers/shared_library/provider_api.h" ?

File: onnxruntime/core/providers/rocm/rocm_execution_provider.cc [hunk 5]
```diff
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 #include "core/providers/shared_library/provider_api.h"
-#include <core/platform/env_var_utils.h>
+#include "core/platform/env_var_utils.h"
```
Comment (zhangyaobit) [inline onnxruntime/core/providers/rocm/rocm_execution_provider.cc:5]: Same here?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 4 above)
Comment (cloudhan) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:6]: Impossible!

`provider_api.h` is the evil header. It triggers ODR violation if not ordered correctly.

File: onnxruntime/python/onnxruntime_pybind_state_common.cc (hunk 2 above)
Comment (cloudhan) [reply onnxruntime/python/onnxruntime_pybind_state_common.cc:27]: Resolved in Teams Chat

File: onnxruntime/core/platform/env_var_utils.h [hunk 6]
```diff
@@ -85,6 +78,14 @@ std::optional<T> ParseTestOnlyEnvironmentVariable(const std::string& name,
     ORT_THROW("Value of environment variable ", name," must be ", oss.str(), ", but got ", raw_env);
   }
 
+  auto env = onnxruntime::ParseEnvironmentVariable<T>(name);
+
+  std::string default_hint = "End users should opt for provider options or session options and must not rely on it.";
```
Comment (zhangyaobit) [inline onnxruntime/core/platform/env_var_utils.h:83]: "End users should opt for provider options or session options"

**Review by zhangyaobit — APPROVED** [review]

**Pull request merged**